OBSDOCS-552:  sample service config for Cluster Observability Operator tech preview (#67608)
openshift/openshift-docs: modules/monitoring-specifying-how-a-service-is-monitored-for-cluster-observability-operator.adoc

// Module included in the following assemblies:
//
// monitoring/cluster-observability-operator/configuring-the-cluster-observability-operator-to-monitor-a-service.adoc

:_mod-docs-content-type: PROCEDURE
[id="specifying-how-a-service-is-monitored-for-cluster-observability-operator_{context}"]

> = Specifying how a service is monitored for Cluster Observability Operator

Author: Changed

> To use the metrics exposed by the sample service you created in the "Creating a `MonitoringStack` object for the Cluster Observability Operator" section, you must configure monitoring components to scrape metrics from the `/metrics` endpoint. 

Author: Changed the reference, but unfortunately it's not possible to add a hyperlink because of restrictions in the OCP docs repo.

> To use the metrics exposed by the sample service you created in the "Creating a `MonitoringStack` object for the Cluster Observability Operator" section, you must configure monitoring components to scrape metrics from the `/metrics` endpoint. 

Reviewer: Suggestion: maybe put something like "See [x] in the "Additional resources" section" and then add an .Additional resources after the module? IDK if this is helpful, though

> To use the metrics exposed by the sample service you created in the "Creating a `MonitoringStack` object for the Cluster Observability Operator" section, you must configure monitoring components to scrape metrics from the `/metrics` endpoint. 

Author: To me, it seems very awkward to do that because all of the sections are on this page. Adding Additional resources section with a link just adds complexity and more mental friction, imo.

> To use the metrics exposed by the sample service you created in the "Creating a `MonitoringStack` object for the Cluster Observability Operator" section, you must configure monitoring components to scrape metrics from the `/metrics` endpoint. 

Author: I think it helps that I include the details about what got configured in the other section in the list of prereqs for this procedure.

You can create this configuration by using a `ServiceMonitor` custom resource definition (CRD) that specifies how the service should be monitored, or a `PodMonitor` CRD that specifies how a pod should be monitored. 

> The `ServiceMonitor` object requires a `Service` object. The `PodMonitor` object does not, which enables Prometheus to scrape metrics directly from the metrics endpoint exposed by a pod

Reviewer 2: which enables **Prometheus** to scrape metrics directly from the metrics
better change as 
which enables **MonitoringStack** to scrape metrics directly from the metrics

Author: Changed

.Prerequisites

* You have access to the cluster as a user with the `cluster-admin` cluster role or as a user with administrative permissions for the namespace.
* You have installed the Cluster Observability Operator.
* You have deployed the `prometheus-coo-example-app` sample service in the `ns1-coo` project.
+
[NOTE]
====
The `prometheus-coo-example-app` sample service does not support TLS authentication. 
====

.Procedure

. Create a YAML file named `example-coo-app-service-monitor.yaml` that contains the following `ServiceMonitor` object configuration details:
+
[source,yaml]
----
apiVersion: monitoring.rhobs/v1alpha1
kind: ServiceMonitor
metadata:
  labels:
    k8s-app: prometheus-coo-example-monitor
  name: prometheus-coo-example-monitor
  namespace: ns1-coo
spec:
  endpoints:
  - interval: 30s
    port: web
    scheme: http
  selector:
    matchLabels:
      app: prometheus-coo-example-app
----
+
This defines a `ServiceMonitor` resource that scrapes the metrics exposed by the `prometheus-coo-example-app` sample service, which includes the `version` metric.
+
[NOTE]
====

> A `ServiceMonitor` resource in a COO namespace can only discover services in the same namespace. That is, the `namespaceSelector` field of the `ServiceMonitor` resource is always ignored.

Reviewer 2: Not sure where do you get this, there is no such limitation "only discover services in the same namespace"

Author: Removed

====

. Apply the configuration to the cluster by running the following command:
+
[source,terminal]
----
$ oc apply -f example-app-service-monitor.yaml
----

> . Verify that the `ServiceMonitor` resource is running by running the following command and observing the output:

Author: Changed

// Module included in the following assemblies:
//
// monitoring/cluster-observability-operator/configuring-the-cluster-observability-operator-to-monitor-a-service.adoc

:_mod-docs-content-type: PROCEDURE
[id="specifying-how-a-service-is-monitored-for-cluster-observability-operator_{context}"]
= Specifying how a service is monitored for Cluster Observability Operator

> To use the metrics exposed by your sample service, you must configure monitoring components to scrape metrics from the `/metrics` endpoint. 

Reviewer: As before, I'd provide some context here about the sample service by referring to the other procedure maybe

Author: Done

> You can create this configuration by using a `ServiceMonitor` custom resource definition (CRD) that specifies how the service should be monitored, or a `PodMonitor` CRD that specifies how a pod should be monitored. 

Reviewer: I would check this with your eng team, but I would think these are custom resources (CRs) or objects rather than CRDs. Could be wrong, though.

Author: @lihongyan1 @jan--f - Are they CRDs or CRs?

Reviewer 2: Both `ServiceMonitor` and `PodMonitor` are CRDs, here you create CRs based these CRD

> The former requires a `Service` object, while the latter does not, which enables Prometheus to scrape metrics directly from the metrics endpoint exposed by a pod.

Reviewer: ```suggestion
The `ServiceMonitor` object requires a `Service` object, while the `PodMonitor` object does not. This enables Prometheus to scrape metrics directly from the metrics endpoint exposed by a pod.
```

I'd spell this out rather than using latter, just because we don't use words like "respectively" and I feel like this is in the same vein. Also, I'd make this sentence shorter if possible by splitting it.

Author: Good suggestion. Made the sentences shorter and clarified the meaning, I hope,

> This procedure shows how to create a `ServiceMonitor` resource for a sample service named `prometheus-coo-example-app` in the `ns1-coo` namespace.

Reviewer: again, I'd just object consistently > resource

Author: Done

.Prerequisites

* You have access to the cluster as a user with the `cluster-admin` cluster role or as a user with administrative permissions for the namespace.
* You have installed the Cluster Observability Operator.
* You have deployed the `prometheus-coo-example-app` sample service in the `ns1-coo` project.
+
[NOTE]
====
The `prometheus-coo-example-app` sample service does not support TLS authentication.
====

.Procedure

> . Create a YAML file for the `ServiceMonitor` resource configuration. For this example, name the file `example-coo-app-service-monitor.yaml`.
> 
> . Add the following `ServiceMonitor` resource configuration details:

Reviewer: ```suggestion
. Create a YAML file named `example-coo-app-service-monitor.yaml` that contains the following `ServiceMonitor` object configuration details:
```
suggestion

Author: Added

+
[source,yaml]
----
apiVersion: monitoring.rhobs/v1alpha1
kind: ServiceMonitor
metadata:
  labels:
    k8s-app: prometheus-coo-example-monitor
  name: prometheus-coo-example-monitor
  namespace: ns1-coo
spec:
  endpoints:
  - interval: 30s
    port: web
    scheme: http
  selector:
    matchLabels:
      app: prometheus-coo-example-app
----
+
This defines a `ServiceMonitor` resource that scrapes the metrics exposed by the `prometheus-coo-example-app` sample service, which includes the `version` metric.
+
[NOTE]
====
A `ServiceMonitor` resource in a COO namespace can only discover services in the same namespace. That is, the `namespaceSelector` field of the `ServiceMonitor` resource is always ignored.
====

> . Apply the configuration to the cluster:
> +
> [source,terminal]
> ----
> $ oc apply -f example-app-service-monitor.yaml
> ----

Reviewer: ```suggestion
. Apply the configuration to the cluster by running the following command:
+
[source,terminal]
----
$ oc apply -f example-app-service-monitor.yaml
----
```

Author: Added

> . Verify that the `ServiceMonitor` resource is running:
> +
> [source,terminal]
> ----
> $ oc -n ns1-coo get servicemonitor
> ----

Reviewer: ```suggestion
. Verify that the `ServiceMonitor` resource is running, by running the following command and observing the output:
+
[source,terminal]
----
$ oc -n ns1-coo get servicemonitor
----
```

Author: added
